earthfile2llb/interpreter.go

```diff
@@ -972,7 +971,11 @@ func (i *Interpreter) handleCopy(ctx context.Context, cmd spec.Command) error {
 	if !allClassical && !allArtifacts {
 		return i.errorf(cmd.SourceLocation, "combining artifacts and build context arguments in a single COPY command is not allowed: %v", srcs)
 	}
-
+	for _, src := range srcs {
+		if src == "" {
+			return i.wrapError(err, cmd.SourceLocation, "empty COPY source encountered")
```

Comment:
```suggestion
			return i.wrapError(err, cmd.SourceLocation, "empty COPY source encountered")
			return i.errorf(cmd.SourceLocation, "COPY command is referencing an empty src path (maybe an $ARG is missing or set to \"\")")
```

Reply:
I'm actually not sure we should necessarily error in this case (I originally worked on this ticket and meant to get back to it) - I _think_ technically (not sure of the use case though) the arg could be intentionally left empty.
Perhaps printing a warning instead? thoughts?

Reply:
For `COPY`, specifically, I think we'll need a non-empty file path.

Reply:
it's not possible to have an empty file name under linux:

```
alex@surfperch:~$ touch ""
touch: cannot touch '': No such file or directory
```

however, for the case of a command like `COPY --if-exists "" dst-file` , maybe this should pass, and essentially becomes a non-operation?

Reply:
right, but the srcs slice can contain multiple args, so I can see something like this:
```
my-target:
   ARG file1=foo
   ARG file2
   COPY --if-exists $file1 $file2 some/dest/.
```

`earthly +my-target --file2=bar` => copies foo & bar into some/dest/.

`earthly +my-target` => copies foo (only) into some/dest/.

if we return an error when file2 is empty it would be a breaking change (which I agree can be worked around with an IF statement but still a breaking change).

Perhaps warning when any `src` is empty and error is all `srcs` are empty?

Reply:
great point regarding multiple files!

> Perhaps warning when any src is empty and error is all srcs are empty?

I don't think we need to display a warning, or error in this case -- the user explicitly told us to ignore missing files by setting the `--if-exists` flag.

Reply:
in that case perhaps the simplest way to address this is to skip empty src  in https://github.com/earthly/earthly/blob/96676e71fb7700801d5847f860679bec378d0468/util/llbutil/copyop.go#L32

Reply:
I don't feel strongly about the warning - my only thinking was that there could be a difference between 
 - the arg set but the file does not exist
 - the arg is not set (by mistake).
 and that the user did mean to copy the file (if it exists) but forgot to set the arg for it (so essnetially the file might exist but the arg doesn't point to it).
 
 But like I said, no strong feelings about it.

Reply:
That makes sense. The second point was basically what I was thinking when I wrote this code. An empty source seems erroneous. We can leave that point for later and just use the change you suggested, though.

Reply:
> I see the point, but in what case would '' ever be a valid COPY arg? Is it a missing file, or rather an invocation error? Does it make sense to address both scenarios (empty sources & current panic bug)? Is this a breaking change if we're fixing a preexisting panic? I'm not bound to either approach, just looking to clarify :)

you're right, it would still panic, but this wouldn't:
```
my-target:
   ARG file1=foo //foo exists
   ARG file2
   COPY $file1 $file2 some/dest/.
```
 so maybe skipping/erroring in the copyop.go instead of interperter.go would be better.

Reply:
Good point. I'll stick with ignoring the empty file for now. Thanks for the discussion.

Reply:
side note: what would be great is to be able to know if the used ARG is even defined, but I think that's harder to do (it would require parsing src and finding all the used args in the src expression, then checking if they are all in declared). A problem for another issue I guess.

Reply:
I considered that as well. It does seem odd to expand an undefined arg to `""`. Perhaps this ought to lead to an error instead.

Reply:
> It does seem odd to expand an undefined arg to `""`. Perhaps this ought to lead to an error instead.

I _think_ it's done to mimic the default behaviour of bash, but I agree it can be error-prone.

It would be neat if we could have a `set -u` inspired keyword that could be defined once per Earthfile.